tso-server/src/main/java/com/yahoo/omid/tso/PersistenceProcessorImpl.java

```diff
@@ -15,345 +15,216 @@
 
-        if (!isRetry) {
-            reply.abortResponse(startTimestamp, channel, monCtx);
-            return;
-        }
-
-        // If is a retry, we must check if it is a already committed request abort.
-        // This can happen because a client could have missed the reply, so it
-        // retried the request after a timeout. So we added to the batch and when
-        // it's flushed we'll add events to the retry processor in order to check
-        // for false positive aborts. It needs to be done after the flush in case
-        // the commit has occurred but it hasn't been persisted yet.
-        batch.addUndecidedRetriedRequest(startTimestamp, channel, monCtx);
+        requestProcessor.start(requestExec);
     }
 
-    // no event has been received in the timeout period
     @Override
-    public void onTimeout(final long sequence) {
-        maybeFlushBatch();
-    }
+    public void reset() {
+        batchIDCnt = 0L;
 
-    /**
-     * Flush the current batch if it is larger than the max batch size,
-     * or BATCH_TIMEOUT_MS milliseconds have elapsed since the last flush.
-     */
-    private void maybeFlushBatch() {
-        if (batch.isFull()) {
-            flush();
-        } else if ((System.nanoTime() - lastFlush) > TimeUnit.MILLISECONDS.toNanos(BATCH_TIMEOUT_MS)) {
-            timeoutMeter.mark();
-            flush();
-        }
+        batchPool.reset();
+        batch = batchPool.getNextEmptyBatch();
+
+        reply.reset();
     }
 
     @VisibleForTesting
-    synchronized void flush() {
-        lastFlush = System.nanoTime();
-
-        boolean areWeStillMaster = true;
-        if (!leaseManager.stillInLeasePeriod()) {
-            // The master TSO replica has changed, so we must inform the
-            // clients about it when sending the replies and avoid flushing
-            // the current batch of TXs
-            areWeStillMaster = false;
-            // We need also to clear the data in the buffer
-            writer.clearWriteBuffer();
-            LOG.trace("Replica {} lost mastership before flushing data", tsoHostAndPort);
-        } else {
-            try {
-                writer.flush();
-            } catch (IOException e) {
-                panicker.panic("Error persisting commit batch", e.getCause());
-            }
-            batchSizeHistogram.update(batch.getNumEvents());
-            if (!leaseManager.stillInLeasePeriod()) {
-                // If after flushing this TSO server is not the master
-                // replica we need inform the client about it
-                areWeStillMaster = false;
-                LOG.warn("Replica {} lost mastership after flushing data", tsoHostAndPort);
-            }
-        }
-        flushTimer.update((System.nanoTime() - lastFlush));
-        batch.sendRepliesAndReset(reply, retryProc, areWeStillMaster);
-
+    void setBatch(Batch batch) {
```

Comment:
Better make the batch field package local than add a method explicitly.
`@VisibleForTesting` means that method is really `private` but you want it to be visible from unit tests as well. Smelly practice but OK. Here you have a method that is *only* used in unit tests